File: labs/ch01-python-basics/src/03_classes_and_oop.py
"""03 · 类与面向对象 — Python 的 class 比 Java 简洁但功能更野.

核心对比:
    Java                            Python
    ──────                          ──────
    public class User { ... }       class User: ...
    private int age                 self._age (约定, 不强制)
    @Data (Lombok)                  @dataclass
    public User(int x) {...}        def __init__(self, x: int): ...
    extends Animal                  class Dog(Animal): ...
    @Override                       (没有, 直接定义同名方法即可)
    static int count = 0            class 体内 count: int = 0 (类变量)
    abstract class                  ABC + @abstractmethod
    @Getter / @Setter               @property + @x.setter

Run:
    python 03_classes_and_oop.py
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field


# ===== 1. 经典 class — 手写 __init__ =====
class Animal:
    """动物 — 演示构造、实例方法、特殊方法."""

    # 类变量 (相当于 Java static), 所有实例共享
    species_count: int = 0

    def __init__(self, name: str, age: int) -> None:
        # 实例变量, 加 self.
        self.name = name
        self._age = age  # 单下划线: 约定为"内部使用"
        Animal.species_count += 1

    def speak(self) -> str:
        return f"{self.name} 发出了未知的叫声"

    def __repr__(self) -> str:
        # 对应 Java 的 toString()
        return f"Animal(name={self.name!r}, age={self._age})"

    def __eq__(self, other: object) -> bool:
        # 对应 Java 的 equals()
        if not isinstance(other, Animal):
            return NotImplemented
        return self.name == other.name and self._age == other._age


# ===== 2. 继承 — 和 Java 一样, 但不写 extends =====
class Dog(Animal):
    """子类 — 重写父类方法不需要 @Override."""

    def __init__(self, name: str, age: int, breed: str) -> None:
        super().__init__(name, age)  # 类似 Java 的 super(name, age)
        self.breed = breed

    def speak(self) -> str:
        # 直接定义同名方法即覆盖父类
        return f"{self.name} ({self.breed}) 汪汪叫"


# ===== 3. dataclass — 一行代替 Lombok @Data =====
@dataclass
class User:
    """用 @dataclass 自动生成 __init__/__repr__/__eq__, 比 Lombok 更内置."""

    id: str
    name: str
    age: int = 0
    tags: list[str] = field(default_factory=list)  # 可变默认值必须用 factory


# ===== 4. @property — 对应 Java getter/setter =====
class BankAccount:
    """演示 property: 调用方写 acc.balance 像访问字段, 实际上跑了一个方法."""

    def __init__(self, owner: str, balance: float = 0.0) -> None:
        self.owner = owner
        self._balance = balance

    @property
    def balance(self) -> float:
        """getter — acc.balance 触发这里."""
        return self._balance

    @balance.setter
    def balance(self, value: float) -> None:
        """setter — acc.balance = 100 触发这里."""
        if value < 0:
            raise ValueError("余额不能为负")
        self._balance = value

    def deposit(self, amount: float) -> None:
        self._balance += amount


# ===== 5. 抽象类 — Java 的 abstract class =====
class Shape(ABC):
    """抽象类 — 不能实例化, 子类必须实现 area()."""

    @abstractmethod
    def area(self) -> float:
        """子类必须实现 — 否则实例化会报错."""

    def describe(self) -> str:
        # 普通方法 — 子类可以继承不重写
        return f"{type(self).__name__} 的面积是 {self.area():.2f}"


class Circle(Shape):
    def __init__(self, radius: float) -> None:
        self.radius = radius

    def area(self) -> float:
        return 3.14159 * self.radius * self.radius


class Rectangle(Shape):
    def __init__(self, w: float, h: float) -> None:
        self.w = w
        self.h = h

    def area(self) -> float:
        return self.w * self.h


def main() -> None:
    print("--- 1. Animal & Dog (经典 class + 继承) ---")
    a1 = Animal("Tom", 3)
    d1 = Dog("Buddy", 5, "Golden")
    print(f"a1 = {a1}")
    print(f"d1 = {d1}")
    print(f"a1.speak() = {a1.speak()}")
    print(f"d1.speak() = {d1.speak()}")
    print(f"Animal.species_count = {Animal.species_count}")
    print(f"a1 == Animal('Tom', 3) = {a1 == Animal('Tom', 3)}")

    print("\n--- 2. User (dataclass — Lombok @Data 等价) ---")
    u = User(id="u1", name="Alice", age=30, tags=["admin", "vip"])
    print(f"u = {u}")
    print(f"u.name = {u.name}")

    print("\n--- 3. BankAccount (property 演示 getter/setter) ---")
    acc = BankAccount("Frank", 1000.0)
    print(f"acc.balance = {acc.balance}")  # 看起来访问字段, 其实调用了 getter
    acc.deposit(500)
    print(f"acc.balance after deposit(500) = {acc.balance}")
    try:
        acc.balance = -100
    except ValueError as e:
        print(f"setter 拦截负数: {e}")

    print("\n--- 4. Shape 抽象类 + 多态 ---")
    shapes: list[Shape] = [Circle(5), Rectangle(3, 4)]
    for s in shapes:
        print(s.describe())

    try:
        Shape()  # 抽象类不能实例化
    except TypeError as e:
        print(f"抽象类拒绝实例化: {e}")


if __name__ == "__main__":
    main()

File: labs/ch01-python-basics/src-java/03_ClassesAndOop.java
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * 等价 Python: src/03_classes_and_oop.py — 看这个文件感受 JDK 23 在做和 Python 同样的事.
 *
 * <p>核心对照:
 * <pre>
 *   Java 8 老写法                           JDK 23 新写法                      Python 写法
 *   ─────────────                           ──────────────                    ────────────
 *   POJO + getter/setter (Lombok @Data)     record User(...)                  &#64;dataclass class User
 *   abstract class + 子类                    sealed interface + permits         ABC + &#64;abstractmethod
 *   instanceof + cast                        pattern matching switch            isinstance + duck typing
 *   重写 + &#64;Override                          同左 (但子类隐式继承)               直接定义同名方法
 *   private int age + getAge()              record 自带 accessor               &#64;property
 * </pre>
 *
 * <p>JDK 23 特性使用:
 * <ul>
 *   <li>{@code record} — 替代 Lombok &#64;Data / Python &#64;dataclass</li>
 *   <li>{@code sealed interface} — 限定子类范围 (Python 没有等价, 用 Protocol/ABC 模拟)</li>
 *   <li>Pattern matching switch — 比 instanceof 链更优雅</li>
 *   <li>{@code var} — 局部类型推断</li>
 * </ul>
 */
class ClassesAndOop {

    // ===== 1. 经典 class — 手写 ctor (Python 的 __init__) =====
    /** 动物 — 演示传统 class. JDK 23 仍保留, 但有状态/继承时才用. */
    static class Animal {

        /** 类变量 (Java static field) — 等价 Python class 体内的 species_count. */
        static int speciesCount = 0;

        protected final String name;
        protected final int age;

        Animal(final String name, final int age) {
            this.name = name;
            this.age = age;
            speciesCount++;
        }

        /** Python 直接定义同名方法重写, Java 写 &#64;Override 增强可读性. */
        public String speak() {
            return "%s 发出了未知的叫声".formatted(name);
        }

        /** 等价 Python 的 __repr__. */
        @Override
        public String toString() {
            return "Animal(name=%s, age=%d)".formatted(name, age);
        }

        /** 等价 Python 的 __eq__. */
        @Override
        public boolean equals(final Object other) {
            return other instanceof Animal a && Objects.equals(name, a.name) && age == a.age;
        }

        @Override
        public int hashCode() {
            return Objects.hash(name, age);
        }
    }

    // ===== 2. 继承 — extends 关键字 (Python 是 class Dog(Animal):) =====
    static class Dog extends Animal {

        private final String breed;

        Dog(final String name, final int age, final String breed) {
            super(name, age);
            this.breed = breed;
        }

        @Override
        public String speak() {
            return "%s (%s) 汪汪叫".formatted(name, breed);
        }
    }

    // ===== 3. record — JDK 14+, 一行干掉 Lombok @Data + Python @dataclass =====
    /**
     * 用户 — record 自动生成 ctor / accessor / equals / hashCode / toString.
     * 等价 Python {@code @dataclass class User: id: str; name: str; age: int = 0}.
     */
    record User(String id, String name, int age, List<String> tags) {
        /** Compact constructor — 校验 + 防御性拷贝 (Python @dataclass 也支持 __post_init__). */
        public User {
            Objects.requireNonNull(id, "id 不能为空");
            tags = List.copyOf(tags); // 不可变拷贝, 防止外部篡改
        }

        /** 简化 ctor — Java 没有"默认参数", 用 overload 模拟 Python age=0, tags=[]. */
        public User(final String id, final String name) {
            this(id, name, 0, List.of());
        }
    }

    // ===== 4. record 也能做 getter (没 setter, record 是不可变的) =====
    /** 银行账户 — 演示 record 不可变 + 演化方法返回新实例 (函数式风格). */
    record BankAccount(String owner, double balance) {

        public BankAccount {
            if (balance < 0) {
                throw new IllegalArgumentException("余额不能为负");
            }
        }

        /** 存款返回新 record (Python @property setter 等价是直接 mutate, 这里更函数式). */
        public BankAccount deposit(final double amount) {
            return new BankAccount(owner, balance + amount);
        }
    }

    // ===== 5. sealed interface — JDK 17+, 比 Python ABC 更严格 (限定子类) =====
    /**
     * 形状 — sealed interface 限定只有 Circle/Rectangle 实现.
     * Python 的 ABC 只能限定"必须实现哪些方法", 限定不了"谁能继承".
     */
    sealed interface Shape permits Circle, Rectangle {

        double area();

        default String describe() {
            return "%s 的面积是 %.2f".formatted(this.getClass().getSimpleName(), area());
        }
    }

    record Circle(double radius) implements Shape {
        @Override
        public double area() {
            return Math.PI * radius * radius;
        }
    }

    record Rectangle(double w, double h) implements Shape {
        @Override
        public double area() {
            return w * h;
        }
    }

    // ===== 6. Pattern matching switch on sealed type — JDK 21+ stable =====
    /** 编译器知道 sealed 的所有子类, 不写 default 都不会警告 (穷尽性检查). */
    static String classify(final Shape shape) {
        return switch (shape) {
            case Circle c -> "圆形 半径=%.2f".formatted(c.radius());
            case Rectangle r -> "矩形 %.2fx%.2f".formatted(r.w(), r.h());
        };
    }

    public static void main(final String[] args) {
        System.out.println("--- 1. Animal & Dog (经典 class + 继承) ---");
        final var a1 = new Animal("Tom", 3);
        final var d1 = new Dog("Buddy", 5, "Golden");
        System.out.println("a1 = " + a1);
        System.out.println("d1 = " + d1);
        System.out.println("a1.speak() = " + a1.speak());
        System.out.println("d1.speak() = " + d1.speak());
        System.out.println("Animal.speciesCount = " + Animal.speciesCount);
        System.out.println("a1.equals(new Animal('Tom', 3)) = " + a1.equals(new Animal("Tom", 3)));

        System.out.println("\n--- 2. User (record — Lombok @Data 等价) ---");
        final var tags = new ArrayList<String>();
        tags.add("admin");
        tags.add("vip");
        final var u = new User("u1", "Alice", 30, tags);
        System.out.println("u = " + u);
        System.out.println("u.name() = " + u.name());

        System.out.println("\n--- 3. BankAccount (record 不可变 + 函数式演化) ---");
        var acc = new BankAccount("Frank", 1000.0);
        System.out.println("acc.balance() = " + acc.balance());
        acc = acc.deposit(500); // 返回新实例
        System.out.println("acc.balance() after deposit(500) = " + acc.balance());
        try {
            new BankAccount("Frank", -100);
        } catch (IllegalArgumentException e) {
            System.out.println("compact ctor 拦截负数: " + e.getMessage());
        }

        System.out.println("\n--- 4. Shape sealed + pattern matching ---");
        final List<Shape> shapes = List.of(new Circle(5), new Rectangle(3, 4));
        for (final var s : shapes) {
            System.out.println(s.describe() + " | " + classify(s));
        }
    }
}
